### PR TITLE
openspec-review [SKILL]

### PR DESCRIPTION
**Pull request by iromu**
Iterative multi-agent loop that reviews, critiques, and implements changes until the OpenSpec is fully satisfied and all tests pass. Use when the user wants a full review–fix cycle on one or more OpenSpec changes before archiving.

Runs an iterative review–critique–implement loop on an OpenSpec change: spawn subagents from the self-contained role briefs in the skill, apply fixes where issues are found, and repeat until the consensus check passes.

Related issue #1741 

Alternative to PR #1752 

## Summary by CodeRabbit

- **New Features**
  - Added an optional Review workflow for iteratively reviewing, critiquing, fixing, and validating OpenSpec changes.
  - Added the `/opsx:review` command and `openspec-review` skill.
  - Review completion requires clean reviewer and critic results, successful validation, and passing tests or builds.
  - Added support for selecting changes and optionally scoping reviews to a specific store.

- **Documentation**
  - Updated workflow lists, setup guidance, skills references, and the glossary to document the new Review workflow.

### CODE REVIEW

**Comment by iromu** [comment]
I run this after "apply" and before "archiving" to validate that everything is implemented as intended. The goal is to make local LLM models more effective when using Openspec, providing an additional layer of reasoning and critique on the applied changes. 

A thinking model may capture and detect everything this skill intents to intercept. But this gate ensures another check for quality.

File: openspec/changes/archive/2026-09-06-add-review-workflow/design.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+### D4: Frontmatter is generator-shaped, not hand-shaped
+`generateSkillContent()` emits the frontmatter; the template only supplies `name`, `description`, `license`, `compatibility`, `metadata`. The committed copy already matches the generator's shape once `stripVolatileFrontmatter` removes `generatedBy`, so no frontmatter special case is needed. The `allowed-tools: Bash(openspec:*)` value stays as in every sibling skill: it is house boilerplate for the openspec CLI surface, while the write-capable IMPLEMENTER/SIMPLIFIER roles are governed by the prose boundaries inside the role briefs themselves (read-only roles never write; fixers are surgical and test-gated). A per-role tool-gating mechanism would be a separate change.
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/design.md:39]: The enforcement gap was real: prose inside a brief does not stop a tool call. The template now states that the executor enforces the boundary with capabilities rather than instruction alone — the read-only roles are spawned without file-editing or install tools, their shell is kept to read-only inspection such as `git diff`, and the build command stays with the executor. See the `Read-only roles never write` guardrail in `src/core/templates/workflows/review.ts` (both the skill and the command literal), commit 7c45351, and the matching `Reviewer write boundary` scenario in `openspec/specs/opsx-review-skill/spec.md`.

D4 itself is left as written. It records the decision taken while authoring the change, and the hard per-role tool gate it deferred is still a separate piece of work; what landed here is the executor's obligation to restrict capabilities when it spawns a read-only role.

File: openspec/changes/archive/2026-09-06-add-review-workflow/design.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+### D5: Remove `docs/openspec-review/SKILL.md` when the template lands
+The reference copy has no home in the docs tree (no page references it) and would drift from the template. Its content survives in git history and in `src/core/templates/workflows/review.ts`. Alternative considered: keep as documentation — rejected, a second unowned copy of generated content is exactly what the parity guard exists to prevent.
+
+### D6: Delta specs make minimal surgical edits to existing enumerations
+The `cli-init` delta adds `openspec-review/SKILL.md` and `/opsx:review` to the two generated-set enumerations and adds one review-availability scenario, keeping the surrounding requirement text as-is. The pre-existing staleness of those enumerations (they list 9 entries while the pipeline today generates 12) is observed, recorded here, and deliberately not addressed in this change.
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/design.md:45]: Fixed in the live spec rather than here: `openspec/specs/cli-init/spec.md` now enumerates all thirteen skill directories and all thirteen slash commands the pipeline emits, and states the core/opt-in split once as a rule instead of asserting a count that was wrong for whichever profile a reader had in mind (commit 951a54f). The lists were checked against their sources of truth — `ALL_WORKFLOWS` and `CORE_WORKFLOWS` in `src/core/profiles.ts`, and `COMMAND_IDS` in `src/core/shared/tool-detection.ts` — and `test/core/init.test.ts` passes green against that description (127 tests).

The archived D6 stays as written: it records what was observed and deliberately deferred at authoring time, and rewriting it would falsify the record of a delivered change.

File: openspec/changes/archive/2026-09-06-add-review-workflow/specs/cli-init/spec.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
+## MODIFIED Requirements
+
+### Requirement: Skill Generation
+
+The command SHALL generate Agent Skills for selected AI tools.
+
+#### Scenario: Generating skills for a tool
+
+- **WHEN** a tool is selected during initialization
+- **THEN** create 10 skill directories under `.<tool>/skills/`:
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/specs/cli-init/spec.md:10]: Applied to the live canonical spec, which is the copy the workflow reads: `openspec/specs/cli-init/spec.md` now derives both artifact sets from the effective workflow selection instead of requiring ten of each, so the `review` workflow can be opt-in (D2) without contradicting the enumeration (commits 7c45351 and 951a54f). The `Review skill installed for a selected workflow set` scenario still carries the negative case for the default core profile.

The delta copy under `openspec/changes/archive/` is the record of what that change proposed; it is left unmodified on purpose.

File: openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+- **AND** reads every artifact named by the status output (proposal, design, delta specs, tasks)
+- **AND** the context block names the change, artifact paths, affected modules/paths, and the build and test commands for those modules
+
+### Requirement: Reviewer Role
+The skill SHALL instruct a read-only reviewer to audit the implementation against the change artifacts and report findings with evidence.
+
+#### Scenario: Reviewer audit
+- **WHEN** the reviewer role is run with the change context block
+- **THEN** the reviewer reads every artifact, verifies each delta-spec requirement and scenario against code citing `file:line` evidence, and verifies each task checkbox is backed by implementation and tests
+- **AND** runs the build and test commands from the context block and reports failures with the relevant output
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md:39]: Agreed — an implementation-plus-test rule is the wrong yardstick for a documentation or CI task, and this change's own task list contains several of those. The reviewer brief and the consensus condition now key the evidence to the task's type: a code task needs the implementation plus its test; a documentation, coordination, or CI task needs the outcome it states. A checkbox with no evidence of the right kind is still a must-fix finding. See `src/core/templates/workflows/review.ts` (both literals), commit 7c45351, and the `Reviewer audit` scenario in `openspec/specs/opsx-review-skill/spec.md`.

File: openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+The skill SHALL end the loop only when all four consensus conditions hold on evidence, and SHALL otherwise feed the failing evidence back into the fix step.
+
+#### Scenario: Consensus reached
+- **WHEN** the reviewer reports `Verdict: CLEAN` and the critic reports `Verdict: PASS` and `openspec validate --change "<name>" --json` reports the change valid and all tests for the affected modules pass with the build completing without errors
+- **THEN** the loop ends and the skill reports consensus REACHED
+
+#### Scenario: Consensus not reached
+- **WHEN** any consensus condition fails
+- **THEN** the failing evidence is fed back into the fix step and the loop repeats
+- **AND** a single clean reviewer pass without the other conditions is not sufficient to end the loop
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md:89]: Step 6 now carries the stop condition the loop was missing: at most 3 cycles per change, and the loop also ends early when a cycle applies no fix or repeats the previous cycle's blocking findings, reporting `NOT REACHED` together with the outstanding evidence. Reinforced by the `The loop is bounded` guardrail in `src/core/templates/workflows/review.ts` and the `The loop is bounded` scenario in `openspec/specs/opsx-review-skill/spec.md` (commit 7c45351). The Consensus Gate requirement was widened to say the failing evidence is fed back "within a bounded number of cycles", so the bounded exit is part of the contract rather than an implementation courtesy.

File: openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+#### Scenario: Pre-existing failures do not block
+- **WHEN** a test or build failure is shown to predate the change
+- **THEN** it is noted separately in the report and does not block consensus
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md:93]: "With proof" was doing a lot of unearned work here, so the exception now names the proof: a failure is excused from consensus only when the same failure is reproduced on the pre-change baseline — a clean checkout of the base commit, or the executor's run recorded before any fix — and that reproduction is quoted in the report. Without it the failure stays blocking. See the `Distinguish pre-existing failures on proof` guardrail in `src/core/templates/workflows/review.ts` and the `Pre-existing failures do not block` scenario in `openspec/specs/opsx-review-skill/spec.md` (commit 7c45351).

File: openspec/changes/archive/2026-09-06-add-review-workflow/specs/cli-init/spec.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/specs/cli-init/spec.md:10]: Both live surfaces now agree with the generator. `openspec/specs/cli-init/spec.md` derives the artifact set from the effective workflow selection, enumerates all thirteen skill/command pairs, and states the core/opt-in rule once (commits 7c45351 and 951a54f); `review` stays opt-in per D2, and the `Review skill installed for a selected workflow set` scenario covers the default-core case where it is absent. Verified against `ALL_WORKFLOWS`/`CORE_WORKFLOWS` in `src/core/profiles.ts` and `COMMAND_IDS` in `src/core/shared/tool-detection.ts`, with `test/core/init.test.ts` green.

The archived delta copies are left as delivered.

File: openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+### Requirement: Critic Role
+The skill SHALL instruct a read-only critic to challenge each reviewer finding with evidence and to search for missed defects.
+
+#### Scenario: Critic challenge
+- **WHEN** the critic role is run with the change context block and the reviewer's findings
+- **THEN** the critic marks each reviewer finding `CONFIRMED` or `REJECTED` with evidence (diff, test output, or quoted code)
+- **AND** reports missed edge cases, logic gaps, silent failures, regressions, scope creep, and over-engineering
+- **AND** classifies every finding as `BLOCKING`, `WARNING`, or `SUGGESTION` with a concrete alternative for each
+- **AND** ends with `Verdict: PASS`, `Verdict: WARNING`, or `Verdict: BLOCKING (N blocking, M warnings, K suggestions)`
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md:55]: This was the loop's worst defect: step 6 demanded `PASS` while step 5 only ever dispatches fixers for confirmed must-fix findings, so a critic that filed a `WARNING` could keep the cycle running forever on findings nothing in the loop was allowed to act upon. Made explicit at all three decision points (commit 7c45351):

- Step 6: `WARNING` and `SUGGESTION` findings never keep the loop open; they are collected for the final report.
- Critic brief: `PASS` means no blocking findings whatever the warning and suggestion count, and `WARNING` exists only to bring a non-blocking finding to the user's attention.
- Consensus check: "Critic verdict `PASS`: no blocking findings; `WARNING` and `SUGGESTION` findings are reported, not gatekept".

Fixers still dispatch for confirmed must-fix findings only (step 5); non-blocking findings reach the human through the cycle report rather than through a fixer. The same rules are in `openspec/specs/opsx-review-skill/spec.md` under `Critic challenge` and `The loop is bounded`.

File: openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+#### Scenario: Consensus reached
+- **WHEN** the reviewer reports `Verdict: CLEAN` and the critic reports `Verdict: PASS` and `openspec validate --change "<name>" --json` reports the change valid and all tests for the affected modules pass with the build completing without errors
+- **THEN** the loop ends and the skill reports consensus REACHED
+
+#### Scenario: Consensus not reached
+- **WHEN** any consensus condition fails
+- **THEN** the failing evidence is fed back into the fix step and the loop repeats
+- **AND** a single clean reviewer pass without the other conditions is not sufficient to end the loop
+
```
[the remark this answers is not in the document]
Comment (iromu) [reply openspec/changes/archive/2026-09-06-add-review-workflow/specs/opsx-review-skill/spec.md:90]: Bounded, and in two directions rather than one (commit 7c45351): a cycle ceiling of 3 per change, and a no-progress stop for a cycle that applies no fix or repeats the previous cycle's blocking findings — the second one matters more, since a cycling loop that keeps re-reporting the same blocking finding would have burned the ceiling every time without ever converging. Both are stated in step 6, in the `The loop is bounded` guardrail in `src/core/templates/workflows/review.ts`, and in the `The loop is bounded` scenario of `openspec/specs/opsx-review-skill/spec.md`. A bounded stop is reported as `NOT REACHED` with the outstanding evidence, never dressed up as consensus.
